src/gateway/bridge/shard_runner.rs

```diff
@@ -171,7 +172,14 @@ impl ShardRunner {
 
             if let Some(event) = event {
                 #[cfg(feature = "collector")]
-                self.collectors.lock().expect("poison").retain_mut(|callback| (callback.0)(&event));
+                {
+                    let read_lock = self.collectors.read();
+                    let to_remove: Vec<_> = read_lock.iter().filter(|callback| !callback.0(&event)).cloned().collect();
+                    drop(read_lock);
+                    if !to_remove.is_empty() {
+                        self.collectors.write().retain(|f| !to_remove.contains(f));
```

Comment:
What happens if a collector gets pushed into this Vec between the read lock being dropped and the write lock being acquired?

Reply:
Nothing, the collector will not be run. But the code should not delete "wrong" collectors.

Reply:
Okay, might be worth adding a comment here as it looks mighty suspicious without looking into it.

Reply:
I did that